sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -306,7 +306,12 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     // decimal type), or that the literal `value` is within range `(min, max)`. For these, we
     // optimize by moving the cast to the literal side.
 
-    val newValue = Cast(Literal(value), fromType).eval()
+    val newValue = try {
+      Cast(Literal(value), fromType).eval()
```

Comment:
can we create a non-ANSI cast here? `Cast(Literal(value), fromType, ansiEnabled = false).eval()`